### PR TITLE
Array

### PR DESCRIPTION
**Pull request by JohanMabille**
Implements #101

### CODE REVIEW

File: include/sparrow/array.hpp [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+            throw std::out_of_range(
+                "array::at: index out of range for array of size " + std::to_string(size())
+                + " at index " + std::to_string(i)
+            );
+        }
+        return (*this)[i];
+    }
+
+    inline auto array::operator[](size_type i) const -> const_reference
+    {
```
Comment (Alex-PLACET) [inline include/sparrow/array.hpp:297]: You should add a SPAROW_ASSERT
same for front(), back(),

File: test/test_array.cpp [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    using const_iter_type = array_iterator<true>;
+
+    TEST_CASE("default constructor")
+    {
+        const_iter_type iter;
+    }
+
+    TEST_CASE_TEMPLATE_DEFINE("all", T, all)
+    {
+        SUBCASE("construcator")
```
Comment (Alex-PLACET) [inline test/test_array.cpp:56]:
```suggestion
        SUBCASE("constructor")
```

**Review by Alex-PLACET — CHANGES_REQUESTED** [review]
I just reviewed the code which is not in the #117

File: include/sparrow/array.hpp [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+            m_iter
+        );
+    }
+
+    template <bool IC>
+    bool array_iterator<IC>::less_than(const self_type& rhs) const
+    {
+        if (m_iter.index() != rhs.m_iter.index())
+        {
+            throw std::invalid_argument("array_iterator::less_than: iterators must have the same type");
```
[the remark this answers is not in the document]
Comment (JohanMabille) [reply include/sparrow/array.hpp:219]: Ah I forgot this one.

**Review by Alex-PLACET — APPROVED** [review]

**Pull request merged**